src/libData/BlockChainData/BlockChain.h

```diff
@@ -0,0 +1,160 @@
+#include "libPersistence/BlockStorage.h"
+
+/// Transient storage for DS/Tx/VC blocks.
+template<class T> class BlockChain
+{
+    std::mutex m_mutexBlocks;
+    CircularArray<T> m_blocks;
+
+protected:
+    /// Constructor.
+    BlockChain() { Reset(); }
+
+    virtual T GetBlockFromPersistentStorage(
+        const boost::multiprecision::uint256_t& blockNum)
+        = 0;
+
+public:
+    /// Destructor.
+    ~BlockChain() {}
+
+    /// Reset
+    void Reset() { m_blocks.resize(BLOCKCHAIN_SIZE); }
+
+    /// Returns the number of blocks.
+    boost::multiprecision::uint256_t GetBlockCount()
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+        return m_blocks.size();
+    }
+
+    /// Returns the last stored block.
+    T GetLastBlock()
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+        try
+        {
+            return m_blocks.back();
+        }
+        catch (...)
+        {
+            return T();
+        }
+    }
+
+    /// Returns the block at the specified block number.
+    T GetBlock(const boost::multiprecision::uint256_t& blockNum)
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+
+        if (blockNum >= m_blocks.size())
+        {
+            LOG_GENERAL(WARNING, "Block number " << blockNum << " absent");
+            return T();
+        }
+        else if (blockNum + m_blocks.capacity() < m_blocks.size())
+        {
+            return GetBlockFromPersistentStorage(blockNum);
+        }
+
+        // To-do: We cannot even index into a vector using uint256_t
+        // uint256_t might just be too big to begin with
+        // Consider switching to uint64_t
+        // For now we directly cast to uint64_t
+
+        if (m_blocks[blockNum].GetHeader().GetBlockNum() != blockNum)
+        {
+            LOG_GENERAL(FATAL,
+                        "assertion failed (" << __FILE__ << ":" << __LINE__
+                                             << ": " << __FUNCTION__ << ")");
+        }
+
+        return m_blocks[blockNum];
+    }
+
+    /// Adds a block to the chain.
+    int AddBlock(const T& block)
+    {
+        boost::multiprecision::uint256_t blockNumOfNewBlock
+            = block.GetHeader().GetBlockNum();
+
+        lock_guard<mutex> g(m_mutexBlocks);
+
+        boost::multiprecision::uint256_t blockNumOfExistingBlock
+            = m_blocks[blockNumOfNewBlock].GetHeader().GetBlockNum();
+
+        if (blockNumOfExistingBlock < blockNumOfNewBlock
+            || blockNumOfExistingBlock == (boost::multiprecision::uint256_t)-1)
+        {
+            m_blocks.insert_new(blockNumOfNewBlock, block);
+        }
+        else
+        {
+            return -1;
+        }
+
+        return 1;
+    }
+};
+
+class DSBlockChain : public BlockChain<DSBlock>
+{
+public:
+    DSBlock GetBlockFromPersistentStorage(
+        const boost::multiprecision::uint256_t& blockNum)
+    {
+        DSBlockSharedPtr block;
+        BlockStorage::GetBlockStorage().GetDSBlock(blockNum, block);
+        return *block;
```

Comment:
Seems we are doing stuff unnecessary here and it's an existing issue in the code. The [`GetDSBlock()`](https://github.com/Zilliqa/Zilliqa/blob/c2b49315a37f36b035d1576ced796b728cc6ca26/src/libPersistence/BlockStorage.cpp#L152) function probably needs some re-design where the return value can be a shared_ptr or unique_ptr. As for the case of not able to get an block, a nullptr wrapped by either share_ptr or unique_ptr can be returned.